File: app/src/main/java/it/manzolo/bluetoothwatcher/updater/AppReceiveSettings.java
package it.manzolo.bluetoothwatcher.updater;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;
import android.util.Log;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;

import it.manzolo.bluetoothwatcher.enums.BluetoothEvents;
import it.manzolo.bluetoothwatcher.enums.WebserviceEvents;
import it.manzolo.bluetoothwatcher.utils.HttpUtils;

public class AppReceiveSettings {

    private static final String TAG = "AppReceiveSettings";

    private final Context context;
    private final String webserviceUrl;
    private final String webserviceUsername;
    private final String webservicePassword;

    public AppReceiveSettings(Context context, String webserviceUrl, String webserviceUsername, String webservicePassword) {
        this.context = context;
        this.webserviceUrl = webserviceUrl;
        this.webserviceUsername = webserviceUsername;
        this.webservicePassword = webservicePassword;
    }

    private String httpGet() throws IOException, JSONException {
        URL url = new URL(this.webserviceUrl + HttpUtils.getSettingsUrl);

        // 1. create HttpURLConnection
        HttpURLConnection loginConn = HttpUtils.loginWebservice(this.webserviceUrl, this.webserviceUsername, this.webservicePassword);

        if (loginConn.getResponseCode() >= 200 && loginConn.getResponseCode() < 400) {
            JSONObject tokenObject = new JSONObject(new HttpUtils().convertStreamToString(loginConn.getInputStream()));
            String token = tokenObject.getString("token");
            Log.d("TOKEN", token);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setUseCaches(false);
            conn.setAllowUserInteraction(false);
            conn.setConnectTimeout(HttpUtils.connectionTimeout);
            conn.setReadTimeout(HttpUtils.connectionTimeout);
            conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
            conn.setRequestProperty("Authorization", "Bearer " + token);

            JSONObject jsonObject = new JSONObject(new HttpUtils().convertStreamToString(conn.getInputStream()));
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this.context);

            SharedPreferences.Editor editor = preferences.edit();
            try {
                editor.putString("webservice_service_every_seconds", jsonObject.get("webservice_service_every_seconds").toString());
            } catch (JSONException e) {
                Log.w(TAG, "webservice_service_every_seconds setting not found");
            }
            try {
                editor.putString("bluetooth_service_every_seconds", jsonObject.get("bluetooth_service_every_seconds").toString());
            } catch (JSONException e) {
                Log.w(TAG, "bluetooth_service_every_seconds setting not found");
            }
            try {
                editor.putString("location_service_every_seconds", jsonObject.get("location_service_every_seconds").toString());
            } catch (JSONException e) {
                Log.w(TAG, "location_service_every_seconds setting not found");
            }
            try {
                editor.putString("update_service_every_seconds", jsonObject.get("update_service_every_seconds").toString());
            } catch (JSONException e) {
                Log.w(TAG, "update_service_every_seconds setting not found");
            }
            try {
                editor.putString("restart_app_service_every_seconds", jsonObject.get("restart_app_service_every_seconds").toString());
            } catch (JSONException e) {
                Log.w(TAG, "restart_app_service_every_seconds setting not found");
            }
            try {
                editor.putString("devices", jsonObject.get("devices").toString());
            } catch (JSONException e) {
                Log.w(TAG, "devices setting not found");
            }
            try {
                editor.putBoolean("enabled", jsonObject.get("enabled").toString().equals("1"));
            } catch (JSONException e) {
                Log.w(TAG, "enabled service setting not found");
            }

            editor.apply();

            Log.d(TAG, "Settings updated");
            return conn.getResponseMessage() + "";
        } else {
            Intent intentWs = new Intent(WebserviceEvents.ERROR);
            intentWs.putExtra("message", "Server login response: " + loginConn.getResponseCode() + " " + loginConn.getResponseMessage());
            LocalBroadcastManager.getInstance(context).sendBroadcast(intentWs);
        }
        Log.e(TAG, "Unable to update settings");
        return "";
    }

    public void receive() {
        new HTTPAsyncTask().execute();
    }

    private JSONObject buidLoginJsonObject() throws JSONException {

        JSONObject jsonObject = new JSONObject();

        jsonObject.put("username", this.webserviceUsername);
        jsonObject.put("password", this.webservicePassword);
        return jsonObject;
    }

    class HTTPAsyncTask extends AsyncTask<String, Void, String> {
        @Override
        protected String doInBackground(String... urls) {
            // params comes from the execute() call: params[0] is the url.
            try {
                try {
                    return httpGet();
                } catch (JSONException e) {
                    Log.e(TAG, e.getMessage());
                    //e.printStackTrace();
                    return "Error: " + e.getMessage();
                }
            } catch (IOException e) {
                //e.printStackTrace();
                Log.e(TAG, e.getMessage());
                Intent intent = new Intent(BluetoothEvents.ERROR);
                // You can also include some extra data.
                intent.putExtra("message", e.getMessage());
                LocalBroadcastManager.getInstance(context).sendBroadcast(intent);
                return "Unable to retrieve web page: " + e.getMessage();
            }
        }

        // onPostExecute displays the results of the AsyncTask.
        @Override
        protected void onPostExecute(String result) {
            Log.d(TAG, "Webserver response: " + result);
        }
    }


}

File: app/src/main/java/it/manzolo/bluetoothwatcher/updater/AppReceiveSettings.kt
package it.manzolo.bluetoothwatcher.updater

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.preference.PreferenceManager
import it.manzolo.bluetoothwatcher.enums.BluetoothEvents
import it.manzolo.bluetoothwatcher.enums.WebserviceEvents
import it.manzolo.bluetoothwatcher.utils.Http
import it.manzolo.bluetoothwatcher.webservice.WebServiceParameters
import kotlinx.coroutines.*
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import kotlin.coroutines.CoroutineContext

class AppReceiveSettings(
    private val context: Context,
    private val webServiceParameter: WebServiceParameters
) : CoroutineScope {
    companion object {
        val TAG: String = AppReceiveSettings::class.java.simpleName

    }

    private var job: Job = Job()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job // to run code in Main(UI) Thread

    fun execute() = launch {
        //onPreExecute()
        val result = doInBackground() // runs in background thread without blocking the Main Thread
        Log.d(TAG, "Webserver response: $result")
        //onPostExecute(result)
    }

    @Suppress("BlockingMethodInNonBlockingContext")
    private suspend fun doInBackground(): String =
        withContext(Dispatchers.IO) { // to run code in Background Thread
            try {
                try {
                    // do async work
                    // 1. create HttpURLConnection
                    val loginConn: HttpURLConnection = Http.loginWebservice(webServiceParameter)
                    if (loginConn.responseCode in 200..399) {
                        val tokenObject = JSONObject(Http().streamToString(loginConn.inputStream))
                        val token = tokenObject.getString("token")
                        Log.d("TOKEN", token)
                        val conn =
                            URL(webServiceParameter.getUrl() + Http.getSettingsUrl).openConnection() as HttpURLConnection
                        conn.requestMethod = "GET"
                        conn.useCaches = false
                        conn.allowUserInteraction = false
                        conn.connectTimeout = Http.connectionTimeout
                        conn.readTimeout = Http.connectionTimeout
                        conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
                        conn.setRequestProperty("Authorization", "Bearer $token")
                        val jsonObject = JSONObject(Http().streamToString(conn.inputStream))
                        val preferences = PreferenceManager.getDefaultSharedPreferences(context)
                        val editor = preferences.edit()
                        try {
                            editor.putString(
                                "webserviceServiceEverySeconds",
                                jsonObject["webserviceServiceEverySeconds"].toString()
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "webserviceServiceEverySeconds setting not found")
                        }
                        try {
                            editor.putString(
                                "bluetoothServiceEverySeconds",
                                jsonObject["bluetoothServiceEverySeconds"].toString()
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "bluetoothServiceEverySeconds setting not found")
                        }
                        try {
                            editor.putString(
                                "locationServiceEverySeconds",
                                jsonObject["locationServiceEverySeconds"].toString()
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "locationServiceEverySeconds setting not found")
                        }
                        try {
                            editor.putString(
                                "updateServiceEverySeconds",
                                jsonObject["updateServiceEverySeconds"].toString()
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "updateServiceEverySeconds setting not found")
                        }
                        try {
                            editor.putString(
                                "restartAppServiceEverySeconds",
                                jsonObject["restartAppServiceEverySeconds"].toString()
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "restartAppServiceEverySeconds setting not found")
                        }
                        try {
                            editor.putString("devices", jsonObject["devices"].toString())
                        } catch (e: JSONException) {
                            Log.w(TAG, "devices setting not found")
                        }
                        /*try {
                            editor.putBoolean("enabled", jsonObject["enabled"].toString() == "1")
                        } catch (e: JSONException) {
                            Log.w(TAG, "enabled service setting not found")
                        }*/
                        try {
                            editor.putBoolean(
                                "autoSettingsUpdate",
                                jsonObject["autoSettingsUpdate"].toString() == "1"
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "autoSettingsUpdate service setting not found")
                        }
                        try {
                            editor.putBoolean(
                                "autoAppUpdate",
                                jsonObject["autoAppUpdate"].toString() == "1"
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "autoAppUpdate service setting not found")
                        }
                        try {
                            editor.putBoolean(
                                "autoAppRestart",
                                jsonObject["autoAppRestart"].toString() == "1"
                            )
                        } catch (e: JSONException) {
                            Log.w(TAG, "autoAppRestart service setting not found")
                        }
                        try {
                            editor.putBoolean("debugApp", jsonObject["debugApp"].toString() == "1")
                        } catch (e: JSONException) {
                            Log.w(TAG, "debug service setting not found")
                        }
                        editor.apply()
                        Log.d(TAG, "Settings updated")
                        return@withContext conn.responseMessage + ""
                    } else {
                        val intentWs = Intent(WebserviceEvents.ERROR)
                        intentWs.putExtra(
                            "message",
                            "Server login response: " + loginConn.responseCode + " " + loginConn.responseMessage
                        )
                        context.sendBroadcast(intentWs)
                    }
                    Log.e(TAG, "Unable to update settings")
                    return@withContext ""
                } catch (e: JSONException) {
                    Log.e(TAG, e.message.toString() + " from " + webServiceParameter.getUrl())
                    "Error: " + e.message.toString() + " from " + webServiceParameter.getUrl()
                }
            } catch (e: IOException) {
                //e.printStackTrace();
                Log.e(TAG, e.message.toString())
                val intent = Intent(BluetoothEvents.ERROR)
                // You can also include some extra data.
                intent.putExtra(
                    "message",
                    e.message.toString() + " from " + webServiceParameter.getUrl()
                )
                context.sendBroadcast(intent)
                "Unable to retrieve web page: " + e.message.toString() + " from " + webServiceParameter.getUrl()
            }

        }

}
